### PR TITLE
librados: Add rados_aio_exec to the C API

### PR DESCRIPTION
**Pull request by iain-buclaw-sociomantic**
This should make adding rados_aio_exec to the C API possible.

Signed-off-by: Iain Buclaw <iain.buclaw@sociomantic.com>

### CODE REVIEW

**Comment by iain-buclaw-sociomantic** [comment]
I also have an implementation of `rados_aio_exec`, tests for C and C++ versions, rados python bindings, and tests for the python ready to commit also.

However I first want to make sure that having a separate override is needed in order to call `aio_exec` from C.  Maybe there is a better way to do it, but it looks like `C_aio_Ack::finish` needs `out_buf` set to copy the data back.

**Comment by liewegas** [comment]
This looks right!  Can you add a test to test/librados/aio.cc?

**Comment by iain-buclaw-sociomantic** [comment]
@liewegas - Sure, should I proceed in adding all other commits I've got lined up also?

**Comment by liewegas** [comment]
Only if they're also aio librados additions; otherwise better to put them in a separate PR.

**Comment by iain-buclaw-sociomantic** [comment]
Yes they are, though I'd probably put the python bindings in a separate PR.

**Comment by iain-buclaw-sociomantic** [comment]
Added `rados_aio_exec` function, tests using the hello plugin, and rebased against master.

File: src/test/librados/aio.cc [hunk 1]
```diff
@@ -3266,6 +3266,73 @@ TEST(LibRadosAioEC, StatRemovePP) {
...
+    sem_wait(test_data.m_sem);
+    sem_wait(test_data.m_sem);
+  }
+  ASSERT_EQ(0, my_completion->get_return_value());
+  AioCompletion *my_completion2 = test_data.m_cluster.aio_create_completion(
+	  (void*)&test_data, set_completion_completeEC, set_completion_safeEC);
+  ASSERT_NE(my_completion2, my_completion_null);
+  bufferlist in, out;
+  ASSERT_EQ(0, test_data.m_ioctx.aio_exec("foo", my_completion2,
+					  "hello", "say_hello", in, &out);
```
Comment (liewegas) [inline src/test/librados/aio.cc]: missing close parens

Comment (iain-buclaw-sociomantic) [inline src/test/librados/aio.cc]: Indeed there is, I thought jenkins ran the testsuite.

**Comment by iain-buclaw-sociomantic** [comment]
Added missing closing parenthesis in both aio_exec tests, and rebased the branch.

File: src/test/librados/aio.cc [hunk 2]
```diff
@@ -3699,6 +3699,73 @@ TEST(LibRadosAioEC, StatRemovePP) {
...
+  {
+    TestAlarm alarm;
+    sem_wait(test_data.m_sem);
+    sem_wait(test_data.m_sem);
+  }
+  ASSERT_EQ(0, rados_aio_get_return_value(my_completion));
+  rados_completion_t my_completion2;
+  ASSERT_EQ(0, rados_aio_create_completion((void*)&test_data,
+	      set_completion_completeEC, set_completion_safeEC, &my_completion2));
+  char out[128]
```
Comment (tchaikov) [inline src/test/librados/aio.cc]: add a ";" please.

@iain-buclaw-sociomantic could you try to compile the test and run it locally? you can 
```
make ceph_test_rados_api_aio
ctest -R ceph_test_rados_api_aio // with a running vstart cluster
```

Comment (iain-buclaw-sociomantic) [inline src/test/librados/aio.cc]: Ah, thanks @tchaikov

**Comment by iain-buclaw-sociomantic** [comment]
@tchaikov - Verified that tests pass locally, rebased the branch.

**Comment by liewegas** [comment]
passed a rados run, lgtm

**Comment by dachary** [comment]
jenkins test this please (flake8 error fixed in master)

**Pull request merged**